### DISCUSSION TITLE
Globe light position

### DISCUSSION THREAD

**Discussion by slaviczavik** [maplibre/maplibre-gl-js#5240] [Q&A]
**maplibre-gl-js version**: v5.0.0-pre.9

**browser**: Chrome 131.0.6778.140 (arm64)

### Steps to Trigger Behavior

 1. Set globe projection.
 2. Set light property with `[1.15, 0, 0]` position.
 3. Globe is illuminated differently than extruded objects.

### Link to Demonstration
See [simple example](https://jsbin.com/ratinuzeye/5/edit?html,output) with globe view and city view next to each other with same position. 

### Expected Behavior
Light position should be same for both views.

### Actual Behavior
Looks like light is not correctly positioned for the globe view.

**Comment by HarelM** [comment]
Globe lighting is the light on "null island".
Extrusion lighting doesn't change from location to location and is somewhat "constant".

**Comment by daliborjanak** [comment]
From my test, it looks like that _light_ is visible only with _sky_ and the default is really strange for globe: 


There is  _light_ set to the null island if you set just projection is style:
```
  "projection": {
    "type": "globe"
  },
```
![Image](https://github.com/user-attachments/assets/2b2fb89f-0861-48b0-8d63-a47920dbb8bd)

but if you add also any _sky_ parameter or only an empty object,
```
  "projection": {
    "type": "globe"
  },
  "sky": {},
```
 then it turns on the _light_ with a default that makes  "night effect" on the visible side. Like this image:
![Image](https://github.com/user-attachments/assets/5e78d9ae-228a-4551-bbce-64c68723757d)

So if you have any _sky_ definition style then it is not usable for the globe. You have to set
```
  "projection": {
    "type": "globe"
  },
  "sky": {
    "atmosphere-blend": 0
  },
```
To see the globe without the "night effect". This is IMHO bug
![Image](https://github.com/user-attachments/assets/145f20a8-e470-45f3-b6d8-285756f6905e)

If you set just _light_, then it doesn't work (another problem):
```
  "projection": {
    "type": "globe"
  },
  "light":{
    "anchor": "map",
    "color": "white",
    "intensity": 0.4,
    "position": [1.15,210,30], //default
  },
```
![Image](https://github.com/user-attachments/assets/b40d64e2-f9ae-4aa4-b443-fac3e5699ce6)

But empty _sky_ with _light_ works and the default is weird 
```
  "projection": {
    "type": "globe"
  },
  "sky":{},
  "light":{
    "anchor": "map",
    "color": "white",
    "intensity": 0.4,
    "position": [1.15,210,30], //default
  },
```
![Image](https://github.com/user-attachments/assets/8c9c251c-8128-447c-a1bd-8136f2a257aa)

**Comment by HarelM** [comment]
In order to see the atmosphere `sky` is needed to be set. If you add an empty sky object you'll get the sky default values for the sky properties.
Sky uses the light properties to define the atmosphere visibility.
I suggest playing with it a bit to better understand how this works.

**Comment by kikuchan** [comment]
Hi,
It seems that the axis of rotation of the lighting for globe projection with `anchor: 'map'` is wrong. (tilted 90 degrees?)

I've successfully got the proper lighting position by using following, a rotation correcting, function:
```ts
function lightPosition(lat: number, lng: number) {
  const sin = (d: number) => Math.sin(d * Math.PI / 180);
  const cos = (d: number) => Math.cos(d * Math.PI / 180);
  const acos = (v: number) => Math.acos(v) * 180 / Math.PI;
  const b = acos(cos(lat) * cos(lng));
  const a = acos(sin(lat) / sin(b));
  return [-1, a, b] as [number, number, number];
}
```

with subsolar point of the Sun:
```js
{
  light: {
    anchor: 'map',
    position: lightPosition(subsolar.lat, subsolar.lng)
  }
}
```

The result:
![image](https://github.com/user-attachments/assets/971cc608-0af5-44b0-837d-3c890c326053)

I've also created a plugin to overlay a night shadow even on a 2D map, just FYI:
https://github.com/kikuchan/maplibre-gl-nightlayer

Thanks.
